### DISCUSSION TITLE
VCS로서 Git이 갖는 이점에 대해 알려주세요.

### DISCUSSION THREAD

**Discussion by javascriptbiggosoo** [wanted-9th-3team/frontend-study#39] [프론트 기술 질문 공간]
프로젝트의 **모든 파일을 필요할 때마다 통째로 저장**해두는 방식과 **Git을 이용해** 버전 관리를 하는 방식은 서로 어떤 차이가 있는지, 그리고 거기서 어떤 이점이 생기는지 알려주세요. 
다른 VCS의 방식과 비교해주셔도 됩니다.(예컨대 SVN 등)

**Comment by michoball** [comment]
git 과 같은 버전관리 시스템을 사용하는 이유는 프로젝트 코드의 유지 보수성 때문입니다. 

개발자는 협업으로 프로젝트를 진행하게되는일이 다반사인데 하나의 프로젝트에 여러명이서 작업을 하다보면 많은 부분에서 의도치않게 오류가 발생하게 됩니다. 이때 코드 작성 히스토리를 보고 오류 코드 이전으로 코드를 되돌리거나 코드가 충돌나는 지점만 파악하여 코드를 수정하는 등 일을 하게 해주는 git 과 같은 시스템을 이용하면 보다 수월하게 프로젝트의 코드를 유지하고 보수할 수 있습니다. 

그렇기 때문에 git을 사용해 관리를 하게 합니다. 

svn 은 실제로 사용해본 적은 없지만 git과 다르게 중앙서버에서 소스코드와 커밋 히스토리를 다루기 때문에 히스토리 관리가 쉽지 않다고 알고 있습니다. 
반면 git 은 로컬에서 소스코드와 히스토리를 다룰 수 있기 때문에 중앙 서버의 코드에 문제가 생겨도 로컬의 코드로 복원 할 수도 있다는 장점이 있어서 git을 많이 사용하게 되었다고 알고 있습니다.

**Comment by WooGie911** [comment]
버젼 관리 시스템으로서 Git이 갖는 이점이라고 한다면 한마디로 분산저장입니다.
개발자로서 가장 중요한것은 유지보수성인데 협업을 하는 과정에서 코드 관리하는 것은 매우 중요합니다.
이 과정에서 버젼관리시스템(VCS)가 탄생하였고 그 중 Git은 다른 방식의 SVN과 비교했을때 여러 장점이 있습니다.

1. 여러 저장소에 분산 저장하기 때문에 네트워크가 없는 상태에서 커밋하여 로컬에 저장 할 수 있다. 저장소가 하나일 경우 네트워크를 통해 코드를 유지 관리 해야하여 불편 할 수 있지만 Git의 경우 여러 저장소를 사용하여 네트워크가 없어도 로컬에 저장해 커밋하여 작업을 이어 갈 수 있다.
2. 중앙 저장 방식이 아니라 분산 저장이기 때문에 하나의 저장소에 문제가 생겨도 다른 저장소에서 작업 중인 작업자의 저장소에 접근하여 복구를 할 수 있다.
3. 저장소를 복제하여 두고 다른 브렌치를 생성함으로써 프로젝트를 진행하거나 관리하기에 용이하다.

![image](https://github.com/wanted-9th-3team/frontend-study/assets/32894573/0f692faf-8951-4716-b879-59bbeb6aec97)

**Comment by javascriptbiggosoo** [comment]
프로젝트의 모든 버전의 원본을 저장한다면 작은 프로젝트에서도 버전의 갯수에 따라 감당할 수 없는 용량으로 늘어날 수 있지만 git commit은 각 파일들의 최종 코드 스냅샷만을 담기 때문에 그렇게 부담이 크지 않습니다.


git commit에서는 모든 파일 하나하나가 링크를 갖습니다. 
새로운 커밋을 만들 때 기존 커밋에서 변함이 없는 파일은 이전 커밋의 코드를 담은 링크만 남깁니다.
git에서 버전을 불러올때는 파일마다 (변경사항이 발생한) 이전 커밋을 참조해서 현재 코드 상태를 계산합니다.


대표적인 VCS로는 Git외에도 SVN이 있습니다. 
SVN은 각 커밋이 그 시점의 최종코드가 아닌 이전 커밋과 비교한 ‘변경사항의 모음’을 저장하기 때문에 버전 하나하나의 용량은 git보다 가볍습니다.
대신 특정 버전의 코드를 읽으려면 그 상위 버전을 전부 읽어야하기 때문에 속도가 느립니다.
